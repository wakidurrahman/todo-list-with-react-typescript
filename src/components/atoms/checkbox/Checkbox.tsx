// src/components/atoms/checkbox/Checkbox.tsx
import React from 'react';

type CheckboxProps = {
  checked: boolean;
  onChange: () => void;
  id: string;
  ariaLabel?: string;
  value?: string;
  children?: React.ReactNode;
};

const Checkbox: React.FC<CheckboxProps> = ({
  checked,
  onChange,
  id,
  ariaLabel,
  value,
  children,
}) => (
  <div className="form-check">
    <input
      className="form-check-input mt-2"
      type="checkbox"
      checked={checked}
      onChange={onChange}
      id={id}
      aria-label={ariaLabel}
      value={value}
    />
    <label
      className={`form-check-label w-100 ${checked ? 'text-decoration-line-through' : ''}`}
      htmlFor={id}
    >
      {children}
    </label>
  </div>
);

export default Checkbox;
