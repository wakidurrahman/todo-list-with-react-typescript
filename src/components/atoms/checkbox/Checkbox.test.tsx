import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Checkbox from './Checkbox';

/**
 * Test suite for the Checkbox component covering:
 * - Basic Rendering
 *   - Renders with required props
 *   - Renders with children content
 * - Functionality
 *   - Checked state handling
 *   - onChange event handling
 * - Styling
 *   - Strike-through style when checked
 *   - Normal style when unchecked
 * - Accessibility
 *   - ARIA attributes
 *   - Label associations (htmlFor/id)
 * - Value Handling
 *   - Custom value attribute
 * - Snapshots
 *   - Unchecked state
 *   - Checked state
 *   - Custom attributes
 */
describe('Checkbox Component', () => {
  const defaultProps = {
    id: 'test-checkbox',
    checked: false,
    onChange: jest.fn(),
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  // Basic rendering tests
  it('renders with required props', () => {
    render(<Checkbox {...defaultProps} />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).not.toBeChecked();
  });

  it('renders with children content', () => {
    const childText = 'Checkbox Label';
    render(<Checkbox {...defaultProps}>{childText}</Checkbox>);

    expect(screen.getByLabelText(childText)).toBeInTheDocument();
  });

  // Functionality tests
  it('handles checked state correctly', () => {
    render(<Checkbox {...defaultProps} checked={true} />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeChecked();
  });

  it('calls onChange when clicked', () => {
    const handleChange = jest.fn();
    render(<Checkbox {...defaultProps} onChange={handleChange} />);

    fireEvent.click(screen.getByRole('checkbox'));
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  // Style tests
  it('applies strike-through style to label when checked', () => {
    render(
      <Checkbox {...defaultProps} checked={true}>
        Label Text
      </Checkbox>
    );

    const label = screen.getByText('Label Text');
    expect(label).toHaveClass('text-decoration-line-through');
  });

  it('does not apply strike-through style when unchecked', () => {
    render(
      <Checkbox {...defaultProps} checked={false}>
        Label Text
      </Checkbox>
    );

    const label = screen.getByText('Label Text');
    expect(label).not.toHaveClass('text-decoration-line-through');
  });

  // Accessibility tests
  it('renders with correct accessibility attributes', () => {
    const ariaLabel = 'Custom checkbox label';
    render(<Checkbox {...defaultProps} ariaLabel={ariaLabel} />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toHaveAttribute('aria-label', ariaLabel);
  });

  it('associates label with input using htmlFor', () => {
    render(<Checkbox {...defaultProps}>Label Text</Checkbox>);

    const checkbox = screen.getByRole('checkbox');
    const label = screen.getByText('Label Text');
    expect(label).toHaveAttribute('for', defaultProps.id);
    expect(checkbox).toHaveAttribute('id', defaultProps.id);
  });

  // Value handling test
  it('renders with custom value', () => {
    const value = 'custom-value';
    render(<Checkbox {...defaultProps} value={value} />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toHaveAttribute('value', value);
  });

  // Snapshot tests
  describe('Snapshots', () => {
    it('matches unchecked checkbox snapshot', () => {
      const tree = renderer
        .create(<Checkbox {...defaultProps}>Unchecked Item</Checkbox>)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches checked checkbox snapshot', () => {
      const tree = renderer
        .create(
          <Checkbox {...defaultProps} checked={true}>
            Checked Item
          </Checkbox>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches checkbox with custom attributes snapshot', () => {
      const tree = renderer
        .create(
          <Checkbox
            {...defaultProps}
            ariaLabel="Custom label"
            value="custom-value"
          >
            Custom Checkbox
          </Checkbox>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
