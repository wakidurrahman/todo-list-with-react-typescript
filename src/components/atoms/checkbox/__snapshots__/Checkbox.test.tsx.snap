// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component Snapshots matches checkbox with custom attributes snapshot 1`] = `
<div
  className="form-check"
>
  <input
    aria-label="Custom label"
    checked={false}
    className="form-check-input mt-2"
    id="test-checkbox"
    onChange={[MockFunction]}
    type="checkbox"
    value="custom-value"
  />
  <label
    className="form-check-label w-100 "
    htmlFor="test-checkbox"
  >
    Custom Checkbox
  </label>
</div>
`;

exports[`Checkbox Component Snapshots matches checked checkbox snapshot 1`] = `
<div
  className="form-check"
>
  <input
    checked={true}
    className="form-check-input mt-2"
    id="test-checkbox"
    onChange={[MockFunction]}
    type="checkbox"
  />
  <label
    className="form-check-label w-100 text-decoration-line-through"
    htmlFor="test-checkbox"
  >
    Checked Item
  </label>
</div>
`;

exports[`Checkbox Component Snapshots matches unchecked checkbox snapshot 1`] = `
<div
  className="form-check"
>
  <input
    checked={false}
    className="form-check-input mt-2"
    id="test-checkbox"
    onChange={[MockFunction]}
    type="checkbox"
  />
  <label
    className="form-check-label w-100 "
    htmlFor="test-checkbox"
  >
    Unchecked Item
  </label>
</div>
`;
